### ISSUE TITLE
Feature Request: Partial/Progressive Provider Init

### ISSUE THREAD

**Issue by shinmog** [hashicorp/terraform-plugin-sdk#111]
This is related to hashicorp/terraform#4149.

When partial/progressive configuration is implemented, it would be nice to allow providers to specify a callback function that would do whatever preparation is necessary before trying to use the provider.

Right now, it feels like the only way to get around any deployment init step is the `null` provider; and if partial / progressive configuration is implemented, then some type of callback to let providers specify a function that gets run exactly one time would transform users' `null` provider workarounds into a standard feature provided by provider maintainers.

**Comment by paultyng** [comment]
Until the time that is implemented, its possible to use a lazy client initialization to do this instead, and only process the configuration when a resource is invoked. I did something similar in a [provider I wrote](https://github.com/paultyng/terraform-provider-unifi/blob/523604c5e50881ae02f9d3c5eea4901d317497d0/internal/provider/lazy_client.go#L69) where use a `sync.Once` and just run the configuration logic on any client operation. Its not very clean, and I agree, a call back for this would be nice, but its possible today until this is fully implemented.
